### DISCUSSION TITLE
Uncertainty calculation of Deep GP incorrect?

### DISCUSSION THREAD

**Discussion by fweberling** [cornellius-gp/gpytorch#1862] [Q&A]
Hi everyone,

the Deep GP implementation has very weird predicted variances. I used the implementation of the Multitask 2-layer Deep GP tutorial and looked at regions away from the data points and used a simple scaled RBF kernel instead of the Matern kernel besides that I didn't change anything in the code. Interestingly, the variance of a specific test point close to the training points changed when I increased the window size and the resolution in torch.linspace(0,1,51) to torch.linspace(-3,4,200) which cannot be right. 

Here, an interval of 200 test points from [0,1] is used.  The predicted variance of the test point right in the middle of each image is var[100,:] = [0.1145, 0.1110, 0.1353, 0.1099] where each element corresponds to the specific task:

![image](https://user-images.githubusercontent.com/52998164/146414852-f6b65f32-8726-4382-b4e8-2571d1807b69.png)

Here, an interval of 200 test points from [-3, 4]  is used. The predicted variance of the same test point right in the middle of each image changes to var[100,:] = [0.6471, 0.8840, 0.7339, 0.5928] where each element corresponds to the specific task. This cannot be right.

![image](https://user-images.githubusercontent.com/52998164/146415462-9faff1be-b7dc-4744-a7cf-88b5c4bffcd4.png)

When I now consider the same implementation of the Deep GP but now only with one layer the uncertainties also change as previously stated for the test point in the middle of the plot and the uncertainties away from the data point look very weird. They have a wave-like structure.

![image](https://user-images.githubusercontent.com/52998164/146416071-d4582404-ef07-4411-9bf5-eaee8309113b.png)

var[100,:] = tensor([0.0520, 0.0501, 0.0608, 0.0536])

![image](https://user-images.githubusercontent.com/52998164/146416279-f04b486f-482c-4fc3-bc34-624fb1cf1aad.png)

var[100,:] = tensor([1.1507, 1.1499, 1.1616, 1.1552])

Is this possibly a bug? I have not changed the code of the tutorial. I only adjusted the resolution of the test point and the interval in torch.linspace(0,1, 51).

I would love to get your feedback!
Friedrich

**Comment by gpleiss** [comment]
The output of a DeepGP model is non-deterministic, so it is possible that the different outputs are due to randomness and not due to different model inputs.

**Reply by fweberling** [reply]
Hi Geoff,

I am sorry but I have to get back to this issue. I tried out several different kernels and different toy datasets for the Deep GP implementation of the GPyTorch tutorial and considered the uncertainty plots. The uncertainty estimates have a very weird almost periodic behaviour. Also it seems that the uncertainty does not become smaller close to the training data.

A 1-Layer Deep GP should correspond to a Sparse Variational GP but the plots of their uncertainty estimates differ substantially. Here is the plot of the 1-Layer Deep GP:

![1-Layer Deep GP Confidence sin-44](https://user-images.githubusercontent.com/52998164/149768463-3d32a82a-507d-4ab1-bccd-34fdd9d86a07.png)

These uncertainties are very strange. I would rather expect them to be like the ones from the Sparse Variational GP:

![SVGP sin-44](https://user-images.githubusercontent.com/52998164/149768612-1e0af792-c487-4764-998d-6b83b6e3b1ce.png)

Do you think these uncertainty estimates are correct? I simply used the Deep GP code from the tutorial. To me it seems there is possibly be something wrong in the Deep GP implementation.

Thank you in advance and sorry for bothering you with this again.

Cheers,
Friedrich

**Reply by fweberling** [reply]
[Deep_GPtrial.zip](https://github.com/cornellius-gp/gpytorch/files/7882635/Deep_GPtrial.zip)

This is the Jupyter Notebook I used.

**Reply by fweberling** [reply]
These weird uncertainty predictions of the Deep GP also do not change if I don't use `gpytorch.settings.fast_pred_var()` . So, I guess it won't be due to the computations at test time using the Lanczos algorithm. Massively increasing the number of inducing points also does not help.

**Reply by fweberling** [reply]
I have found the issue here: 

``````
def predict(self, test_x):
        with torch.no_grad():

            # The output of the model is a multitask MVN, where both the data points
            # and the tasks are jointly distributed
            # To compute the marginal predictive NLL of each data point,
            # we will call `to_data_independent_dist`,
            # which removes the data cross-covariance terms from the distribution.
            preds = model.likelihood(model(test_x)).to_data_independent_dist()

        return preds.mean.mean(0), preds.variance.mean(0)
``````

The predict method in the Deep GP model of the GPyTorch tutorial on Deep GPs does not work correctly. Especially, ```.to_data_independent_dist()``` seems to do something wrong.  I suspect it is about some reshaping.

Using this method the uncertainties become correct:

```
def predict(self, test_x):
    with torch.no_grad():

      preds = model.likelihood(model(test_x))

      preds_mean = preds.mean.mean(axis=0)
      preds_var = preds.covariance_matrix.mean(axis=0).diag().reshape(num_tasks, num_tasks * self.train_x_shape[0]).T

    return preds_mean, preds_var

```
![image](https://user-images.githubusercontent.com/52998164/149985883-e2c821ee-b91b-4a2b-b19a-64fc06db300e.png)

**Reply by gpleiss** [reply]
Hmm okay this does in fact seem like a bug on our end. Thank you so much for digging into this. Let's make this into an issue.

**Reply by gpleiss** [reply]
@fweberling this I think should be closed by #2123
